src/main/scala/sigmastate/serialization/ValueSerializer.scala

```diff
@@ -146,13 +163,33 @@ object ValueSerializer extends SigmaSerializerCompanion[Value[SType]] {
     v
   }
 
+  private val SerializedConstantsStartCode = ConstantCode
+
   def serialize(v: Value[SType]): Array[Byte] = {
```

Comment:
@ergomorphic Should we introduce a new API for (de)serialization using placeholders (like `pack`/`unpack`)? I've made it work in `serialize(v)`/`deserialize(bytes)` treating it as an entry point for the whole tree (de)serialization.

Reply:
@greenhat I think it is better we do additional pass, which optimizes the constants away and replaces them with placeholders.

Reply:
@ergomorphic Do you mean introducing an IR node for constant placeholder and for constants array? I'm a bit confused.
I like the idea of ErgoScript top-level object (IR node?). Its serializer can clip constant placeholder store onto the writer/reader.

Reply:
@greenhat I mean we need to have some ErgoScript object which is not value. This object is responsible for top-level format of serialization. I particular it will check whether GZIP is used, how many constants in array, etc.

Reply:
@ergomorphic Let's assume we make:
```
object ErgoScriptSerializer {
  def serialize(script: Value[SType]): Array[Byte] = ???
  def deserialize(bytes: Array[Byte]): Value[SType] = ???
}
```
What do you mean by 

> it is better we do additional pass, which optimizes the constants away and replaces them with placeholders.

Should we replace the constants with placeholders **after** the tree is serialized with `ValueSerializer.serialize`? If not, and we first traverse the tree and replace the constant with placeholders(special value type?) then we need to teach `ValueSerializer.serialize` how to serialize placeholders. 
Both approaches look quite complicated. Why don't we parametrize `ValueSerializer` with a constant serializer (either regular or placeholder-based)? `ErgoScriptSerialize` would run then run `ValueSerializer` with a placeholder-based constant serializer.

Reply:
@greenhat 1) Constant separation is optional optimization, so everything should work without it.
2) This optimization can be done at compile time, so before serializer.
3) Placeholder is a special value type, which is similar to ValDef nodes. 
ValDef nodes also store an index to common subexpression, however many ValDef nodes can refer to the same subtree, whereus placeholders don't and we should have one-to-one correspondence between constants and placeholders.